### ISSUE TITLE
Publish all artifacts on mavenCentral

### ISSUE THREAD

**Issue by bitspittle** [varabyte/kobweb#103]
Right now we're hosting artifacts on Google cloud in central US but we'll have potential users all over the world, so probably maven central is doing a better job at supporting everyone than my setup is

**Comment by sureshg** [comment]
This is something required to use Kobweb in enterprise setup where maven central / it's proxy is the only approved artifacts repo. Any plans to publish it before it reaches 1.0?  I think Kobweb is stable enough to use for internal tools

**Comment by bitspittle** [comment]
Yes, I will publish to the official gradle repository strong before or at 1.0.

Avoiding it for now because it's a pain (my other libraries Kotter and Truthish use maven and it's a bunch of extra steps)

Apologies for the inconvenience until then. Aiming to get to 1.0 sometime this year.

**Comment by sureshg** [comment]
Sorry to bug you again..any update on this? Compose Multiplatform just become stable with 1.9.20

**Comment by bitspittle** [comment]
So apologies here. Gonna still be several months *at least* before Kobweb is 1.0.

Good news is it's finally a straight shot to 1.0 now ([roadmap document for reference](https://docs.google.com/document/d/1n2Jd02yzuxaatpT7gOhEuijzfaSu9UIaQTV2t4EPcPk/preview)). I've been mainly filling in API gaps and adding the final collection of foundational Silk widgets recently. It just has been taking longer than I thought. (I thought each widget would take a few days but they take more like two weeks each, for example)

It's hard to express how frustrating MavenCentral can be to work with it you've never done it. It adds significant friction to the release schedule every time (and on more than one occasion I've had to delay a release of other artifacts I work by 24-48 hours because the servers just 403 at you constantly). 

With my limited resources, I'm going to be putting this off for as long as possible... (Compose Multiplatform going stable is really nice! But not something I was taking into consideration around this issue)

**Comment by sureshg** [comment]
jfyi: Things are much simpler now with new maven central publishing. You can use this new Gradle plugin for automatic publishing - https://github.com/GradleUp/nmcp

**Comment by bitspittle** [comment]
Thanks for the update! That will be helpful for sure.

Note for now the main issue isn't the publishing step, but just generally supporting two repositories, which increases the odds that something can go wrong. I wish I could say I felt more confident about maven central, but the nexus release process still occasionally causes me issues when working with my other project, Kotter. 

For another example, maven central doesn't update for 30 minutes after you release, which will generally add a weird delay to my current release announcement process. Releases are stressful enough without dealing with that! :)

**Comment by sureshg** [comment]
> generally supporting two repositories (Google and Maven Central)

Since Central is the canonical one for kotlin/jvm ecosystem, do we even need google once we start publishing to Central? Also for 30 minutes delay, i think that got improved with new central publishing and i could able to see the new artifacts in few minutes.  Sorry, i am not pushing you to do this, but having thing s on central is much easier in enterprise environments.

**Comment by bitspittle** [comment]
Once we start publishing to Maven Central, Google won't be the main one anymore. With Kotter I now consider maven the main release blocker, although I still publish everything to Google as well.

For me personally Google is much more convenient -- I own all the bits there and in the past I have cleaned up artifacts that got uploaded there incorrectly.

Also note that for Kobweb 1.0 I'll also need to figure out how to host my gradle plugins in the Gradle repository, just to make things a little more confusing and add more moving parts.

**Comment by bitspittle** [comment]
Ha, so funnily enough, just today I tried publishing snapshots for my library Truthish, targeting both Google and Maven Central repositories.

Maven central rejected me with this message:

```
> * What went wrong:
Execution failed for task ':publishKotlinMultiplatformPublicationToSonatypeMavenRepository'.
> Failed to publish publication 'kotlinMultiplatform' to repository 'SonatypeMaven'
   > Could not GET 'https://s01.oss.sonatype.org/content/repositories/snapshots/com/varabyte/truthish/truthish/1.0.2-SNAPSHOT/maven-metadata.xml'. Received status code 401 from server: Content access is protected by token
```

Why did this break? The code hasn't changed for a long time. It's using the same auth setup that I use in Kotter. This code has working for over a year. Did something change? Was this a temporary fluke and I should just try again later? **I don't know!**

I don't enjoy working with maven central :P

**Comment by sureshg** [comment]
Yeah, i think as a nexus user you might have this communication from Nexus

```

Dear Maven Central publisher,

We are making changes to the OSSRH authentication backend. For most users this should be a transparent process, and you should be able to continue to use your existing username and password to connect the Nexus UI. In case you need to update your password, please follow our documentation.

To configure a publisher’s plugin authentication you would need to update your plugin settings to use a user token instead of the Nexus UI username and password login.

```
You should use token now - https://support.sonatype.com/hc/en-us/articles/360049469534-401-Content-access-is-protected-by-token-when-accessing-repositories

**Comment by bitspittle** [comment]
@sureshg Yeah, I had seen that earlier and assumed it didn't apply to me. After all, it says that the process should be transparent and that username / password should still work. Worse, if you follow through with the instructions on setting up the token mention that this is a Sonatype Pro feature only:

https://help.sonatype.com/en/user-tokens.html

Ah well, I guess this is what I'll be doing today, figuring this stuff out. :)

**Comment by CLOVIS-AI** [comment]
As an additional use-case: creating libraries that build off of Kobweb is quite complex right now, because users of such libraries need to do additional setup before being able to consume them (as they can't download Kobweb artifacts in a default project).

**Comment by bitspittle** [comment]
"Good news"... I just started noticing I've been getting charged for hosting my own artifacts.

I guess enough people are using Kobweb at this point I've gone over a download limit I never originally thought possible. Yay success :)

I guess I am now financially motivated to move things over to maven central a bit sooner than I originally expected. I'll at least try to look closer into what is exactly happening and evaluating a move over the maven central it in the next few weeks.

**Comment by bitspittle** [comment]
As of now, all Kobweb 0.20.1 *snapshot* artifacts are being published to the sonatype snapshot repository.

What this means is that at this very moment you can replace any instance of this line throughout your project's Gradle files:
```kotlin
maven("https://us-central1-maven.pkg.dev/varabyte-repos/public")
```

with this line:
```kotlin
maven("https://s01.oss.sonatype.org/content/repositories/snapshots/")
```

> [!TIP]
> Consider limiting the maven snapshot scope to just kobweb artifacts:
> ```kotlin
> maven("https://s01.oss.sonatype.org/content/repositories/snapshots/") {
>   content { includeGroupByRegex("com\\.varabyte\\.kobweb.*") }
> }
> ```

Most projects should have this line in two places: one in `settings.gradle.kts` for plugin artifacts, and one in either your root or site `build.gradle.kts` file for library artifacts. (If you apply the regex tip, you'd put that in both locations).

Additionally, if everything works without a hitch, when 0.20.1 goes stable, Kobweb plugins should be hosted in `gradlePluginPortal()` and all remaining artifacts in  `mavenCentral()`.

I still reserve the right to bail on supporting the Gradle portal plugin and maven central in 0.20.1 if I hit too many obstacles with either repository, so no promises, but I'll try!

**Comment by bitspittle** [comment]
Alright folks, starting with 0.20.1, we've entered a new era: we are now published on Maven Central and the Gradle Plugin Portal. For people who have been waiting for this since 2022.... well, uh, thanks for your patience.

More here: https://github.com/varabyte/kobweb#gradle-configuration

**Comment by CLOVIS-AI** [comment]
Congrats! https://central.sonatype.com/search?q=com.varabyte.kobweb
